File: 프로그래머스/3/43164. 여행경로/여행경로.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

vector<string> answer;
bool finished = false;

void dfs(string current, vector<vector<string>>& tickets, vector<bool>& used, vector<string>& path, int depth) {
    if (depth == tickets.size()) {
        answer = path;
        finished = true;
        return;
    }
    
    for (int i = 0; i < tickets.size(); i++) {
        if (!used[i] && tickets[i][0] == current) {
            used[i] = true;
            path.push_back(tickets[i][1]);
            dfs(tickets[i][1], tickets, used, path, depth + 1);
            if (finished) return; // 정답 찾았으면 바로 return
            used[i] = false;
            path.pop_back();
        }
    }
}

vector<string> solution(vector<vector<string>> tickets) {
    sort(tickets.begin(), tickets.end()); // 알파벳 순서 우선 탐색을 위해 미리 정렬

    vector<bool> used(tickets.size(), false);
    vector<string> path;
    path.push_back("ICN"); // 시작 공항

    dfs("ICN", tickets, used, path, 0);

    return answer;
}

File: 프로그래머스/3/43164. 여행경로/여행경로.py
from collections import deque

def solution(tickets):
    # dict를 사용해 그래프 구성: 출발지 -> [도착지1, 도착지2, ...]
    graph = {}
    for frm, to in tickets:
        if frm in graph:
            graph[frm].append(to)
        else:
            graph[frm] = [to]
    # 알파벳 순으로 작은 경로부터 탐색하기 위해 역순 정렬
    for frm in graph:
        graph[frm].sort(reverse=True)

    route = []  # 최종 경로를 담을 리스트

    def dfs(airport):
        # 현재 공항에서 출발 가능한 모든 티켓 사용
        # graph에 키가 없으면 get으로 빈 리스트를 리턴(
        while graph.get(airport):
            next_dest = graph[airport].pop()
            dfs(next_dest)
        # 더 이상 사용 가능한 티켓이 없으면 방문 기록에 추가
        route.append(airport)

    # ICN에서 출발
    dfs('ICN')
    # 역추적 방식이므로 뒤집어서 반환
    return route[::-1]